File: src/main/java/model/players/HigherTilePlayer.java
package model.players;

import model.Move;
import model.Position;
import model.Tile;
import java.util.LinkedList;
import java.util.List;

public class HigherTilePlayer extends AbstractPlayer {

    public HigherTilePlayer(String _name) {
        super(_name);
    }

    @Override
    public String description() {
        return "This player selects the higher tile to match the head or the tail of the board";
    }

    @Override
    public Move selectTileEmptyBoard() {
        int max = this.getTiles().stream().mapToInt(t -> t.sum()).max().getAsInt();
        Tile t = this.getTiles().stream().filter(a -> a.sum() == max).findFirst().get();
        return new Move(t, Position.HEAD);
    }

    @Override
    public Move selectTileNonEmptyBoard(LinkedList<Tile> board) {
        int first = board.peekFirst().getFirstNumber();
        int last = board.peekLast().getSecondNumber();
        List<Tile> usableTiles = this.getPossibleTiles(first, last);
        int max = usableTiles.stream().mapToInt(t -> t.sum()).max().getAsInt();
        Tile t = usableTiles.stream().filter(a -> a.sum() == max).findFirst().get();
        return new Move(t, t.hasNumber(first) ? Position.HEAD : Position.TAIL);
    }
}

File: src/main/webapp/scripts/players/HigherTilePlayer.js
import AbstractPlayer from "./AbstractPlayer.js";
import Move from "../Move.js";
import Position from "../Position.js";

export default class HigherTilePlayer extends AbstractPlayer {

    constructor(_name) {
        super(_name);
    }

    description() {
        return "This player selects the higher tile to match the head or the tail of the board";
    }

    selectTileEmptyBoard() {
        let max = Math.max(...this.getTiles().map(t => t.sum()));
        let t = this.getTiles().filter(a => a.sum() === max)[0];
        return new Move(t, Position.HEAD);
    }

    selectTileNonEmptyBoard(board) {
        let first = board[0].getFirstNumber();
        let last = board.slice(-1).getSecondNumber();
        let usableTiles = this.getPossibleTiles(first, last);
        let max = Math.max(...usableTiles.map(t => t.sum()));
        let t = usableTiles.filter(a => a.sum() === max)[0];
        return new Move(t, t.hasNumber(first) ? Position.HEAD : Position.TAIL);
    }
}
